src/or/circuitlist.c

```diff
@@ -608,6 +608,30 @@ circuit_get_global_origin_circuit_list(void)
   return global_origin_circuit_list;
 }
 
+/**
+ * Return true if the circuit is a "full-path" circuit.
+ *
+ * This means its desired_route_len is either above DEFAULT_ROUTE_LEN or
+ * it's a Single Onion Service circuit.
+ */
+static int
+circuit_is_full_path(const origin_circuit_t *circ)
+{
+  if (!circ->build_state)
+    return 0;
+
+  if ((circ->rend_data || circ->hs_ident) &&
```

Comment:
Every HS circuits have the identifier attached to them. I think this is enough to cover the intro/rend circuits because of the path length of 1.

The HSDir circuit with 3 hops will be caught by the next condition below it seems.